aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+        buy_in_active_shares(pool, delegator_address, amount);
+
+        event::emit_event(
+            &mut pool.reactivate_stake_events,
+            ReactivateStakeEvent {
+                pool_address,
+                delegator_address,
+                amount,
+            },
+        );
+    }
+
+    /// Withdraw `amount` of owned inactive stake from the delegation pool at `pool_address`.
+    public entry fun withdraw(delegator: &signer, pool_address: address, amount: u64) acquires DelegationPool {
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+        withdraw_internal(borrow_global_mut<DelegationPool>(pool_address), signer::address_of(delegator), amount);
+    }
+
+    #[view]
+    /// Return whether `pending_inactive` stake can be directly withdrawn from
+    /// the delegation pool, implicitly its stake pool, in the special case
+    /// the validator had gone inactive before its lockup expired.
+    public fun can_withdraw_pending_inactive(pool_address: address): bool {
+        stake::get_validator_state(pool_address) == VALIDATOR_STATUS_INACTIVE &&
+        timestamp::now_seconds() >= stake::get_lockup_secs(pool_address)
+    }
+
+    fun withdraw_internal(pool: &mut DelegationPool, delegator_address: address, amount: u64) {
+        // short-circuit if amount to withdraw is 0 so no event is emitted
+        if (amount == 0) { return };
+
+        let pool_address = get_pool_address(pool);
+        let (withdrawal_exists, withdrawal_lockup_cycle) = pending_withdrawal_exists(pool, delegator_address);
+        // exit if no withdrawal or is not inactive and cannot withdraw pending_inactive stake from stake pool
+        if (!(
+            withdrawal_exists &&
+            (withdrawal_lockup_cycle < pool.observed_lockup_cycle || can_withdraw_pending_inactive(pool_address))
+        )) { return };
+
+        amount = redeem_inactive_shares(pool, delegator_address, amount, withdrawal_lockup_cycle);
+
+        let stake_pool_owner = &retrieve_stake_pool_signer(pool);
+        // stake pool will inactivate entire pending_inactive stake at `stake::withdraw` to make it withdrawable
+        // however, bypassing the inactivation of excess stake (inactivated but now withdrawn) ensures
+        // the OLC is not advanced indefinitely on `unlock`-`withdraw` paired calls
+        if (can_withdraw_pending_inactive(pool_address)) {
+            // get excess stake before being entirely inactivated
+            let (_, _, _, pending_inactive) = stake::get_stake(pool_address);
+            if (withdrawal_lockup_cycle == pool.observed_lockup_cycle) {
+                // `amount` less excess if withdrawing pending_inactive stake
+                pending_inactive = pending_inactive - amount
+            };
+            // escape excess stake from inactivation
+            stake::reactivate_stake(stake_pool_owner, pending_inactive);
```

Comment:
Why do we want to reactivate any remaining inactive stake that is not being withdrawn? Can there be a use case when building liquid staking service on top, where multiple different depositors in the liquid staking service need to withdraw different amount separately?

Reply:
The reasons for not inactivating any excess stake would be that:
1. the OLC is advanced when new inactive stake is detected, which can happen indefinitely in case users trigger the inactivation themselves: unlock a small amount and then call withdraw for some fraction of it and repeat
2. as you said, other delegators' pending_inactive stake would be inactivated implicitly. They can always withdraw and add it back unless validator has become active in the meantime

Reply:
1. Shouldn't OLC only advanced if current timestamp > last OLC's recorded lockup expiration? This was the original implementation and made sense. What's the reason for changing this condition?
2. Adding it back adds the stake to pending_active though. Imagine 5 different depositors unlocking 20 APT each via a liquid staking service (to the delegation pool , there's a single delegator - the liquid staking service). In the current implementation, when one deposit withdraws 20 APT, the remaining 80 APT would be restaked, which means the remaining 4 depositors can no longer withdraw. This seems like a bug?